demos/wordcount/src/main/java/com/datatorrent/demos/wordcount/FileWordCount.java

```diff
@@ -0,0 +1,282 @@
+//
+// Since the EOF is received by a single operator, this operator cannot be partitionable
+//@OperatorAnnotation(partitionable = true)
+public class FileWordCount extends BaseOperator
+{
+  private static final Logger LOG = LoggerFactory.getLogger(FileWordCount.class);
+  private static final String GLOBAL = "global";
+
+  // If topN > 0, only data for the topN most frequent words is output; if topN == 0, the
+  // entire frequency map is output
+  //
+  protected int topN;
+
+  // set to true when we get an EOF control tuple
+  protected boolean eof = false;
+
+  // last component of path (i.e. only file name)
+  // incoming value from control tuple
+  protected String fileName;
+
+  // wordMapFile   : {word => frequency} map, current file, all words
+  // wordMapGlobal : {word => frequency} map, global, all words
+  //
+  protected Map<String, WCPair>
+    wordMapFile = new HashMap<>(),
+    wordMapGlobal = new HashMap<>();
+
+  // resultPerFile : singleton list [TopNMap] with per file data; sent on outputPerFile
+  // resultGlobal : singleton list [wordFreqMap] with per file data; sent on outputGlobal
+  //
+  protected transient List<Map<String, Object>> resultPerFile, resultGlobal;
+
+  // singleton map of fileName to sorted list of (word, frequency) pairs
+  protected transient Map<String, Object> resultFileFinal;
+  protected transient List<WCPair> fileFinalList;
+
+  public final transient DefaultInputPort<List<WCPair>> input = new DefaultInputPort<List<WCPair>>()
+  {
+    @Override
+    public void process(List<WCPair> list)
+    {
+      // blend incoming list into wordMapFile and wordMapGlobal
+      for (WCPair pair : list) {
+        final String word = pair.word;
+        WCPair filePair = wordMapFile.get(word);
+        if (null != filePair) {    // word seen previously in current file
+          WCPair globalPair = wordMapGlobal.get(word);    // cannot be null
+          filePair.freq += pair.freq;
+          globalPair.freq += pair.freq;
+          continue;
+        }
+
+        // new word in current file
+        filePair = new WCPair(word, pair.freq);
+        wordMapFile.put(word, filePair);
+
+        // check global map
+        WCPair globalPair = wordMapGlobal.get(word);    // may be null
+        if (null != globalPair) {    // word seen previously
+          globalPair.freq += pair.freq;
+          continue;
+        }
+
+        // word never seen before
+        globalPair = new WCPair(word, pair.freq);
+        wordMapGlobal.put(word, globalPair);
+      }
+    }
+  };
+
+  @InputPortFieldAnnotation(optional = true)
+  public final transient DefaultInputPort<String> control = new DefaultInputPort<String>()
+  {
+    @Override
+    public void process(String msg)
+    {
+      if (msg.isEmpty()) {    // sanity check
+        throw new RuntimeException("Empty file path");
+      }
+      LOG.info("FileWordCount: EOF for {}, topN = {}", msg, topN);
+      fileName = msg;
+      eof = true;
```

Comment:
Multiple files could be processed in a window and multiple eof messages could be sent.